tables.go

```diff
@@ -0,0 +1,315 @@
+type BlobIndex int
+
+// GuidIndex is an index to a #GUID heap.
+type GUIDIndex int
+
+// Assembly is defined in §II.22.2.
+type Assembly struct {
+	HashAlgID      flags.AssemblyHashAlgorithm
+	MajorVersion   uint16
+	MinorVersion   uint16
+	BuildNumber    uint16
+	RevisionNumber uint16
+	Flags          flags.AssemblyFlags
+	PublicKey      BlobIndex
+	Name           string
+	Culture        string
+}
+
+// assemblyOS is defined in §II.22.3.
+// This record should not be emitted into any PE file,
+// but needed temporarily to calculate sizes and offsets for subsequent tables.
+type assemblyOS struct {
+	OSPlatformID   uint32
+	OSMajorVersion uint32
+	OSMinorVersion uint32
+}
+
+// assemblyProcessor is defined in §II.22.4.
+// This record should not be emitted into any PE file,
+// but needed temporarily to calculate sizes and offsets for subsequent tables.
+type assemblyProcessor struct {
+	Processor uint32
+}
+
+// AssemblyRef is defined in §II.22.5.
+type AssemblyRef struct {
+	MajorVersion     uint16
+	MinorVersion     uint16
+	BuildNumber      uint16
+	RevisionNumber   uint16
+	Flags            flags.AssemblyFlags
+	PublicKeyOrToken BlobIndex
+	Name             string
+	Culture          string
+	HashValue        BlobIndex
+}
+
+// assemblyRefOS is defined in §II.22.6.
+// This record should not be emitted into any PE file,
+// but needed temporarily to calculate sizes and offsets for subsequent tables.
+type assemblyRefOS struct {
+	OSPlatformID   uint32
+	OSMajorVersion uint32
+	OSMinorVersion uint32
+	AssemblyRef    Index[AssemblyRef]
+}
+
+// assemblyRefProcessor is defined in §II.22.7.
+// This record should not be emitted into any PE file,
+// but needed temporarily to calculate sizes and offsets for subsequent tables.
+type assemblyRefProcessor struct {
+	Processor   uint32
+	AssemblyRef Index[AssemblyRef]
+}
+
+// ClassLayout is defined in §II.22.8.
+type ClassLayout struct {
+	PackingSize uint16
+	ClassSize   uint32
+	AssemblyRef Index[TypeDef]
+}
+
+// Constant is defined in §II.22.9.
+type Constant struct {
+	Type    flags.ElementType
+	Padding byte       // 1-byte padding zero
+	Parent  CodedIndex `winmd:"HasConstant"`
+	Value   BlobIndex
+}
+
+// CustomAttribute is defined in §II.22.10.
+type CustomAttribute struct {
+	Parent CodedIndex `winmd:"HasCustomAttribute"`
+	Type   CodedIndex `winmd:"CustomAttributeType"`
+	Value  BlobIndex
+}
+
+// DeclSecurity is defined in §II.22.11.
+type DeclSecurity struct {
+	Action        uint16
+	Parent        CodedIndex `winmd:"HasDeclSecurity"`
+	PermissionSet BlobIndex
+}
+
+// EventMap is defined in §II.22.12.
+type EventMap struct {
+	Parent    Index[TypeDef]
+	EventList Slice[Event]
+}
+
+// Event is defined in §II.22.13.
+type Event struct {
+	EventFlags flags.EventAttributes
+	Name       string
+	EventType  CodedIndex `winmd:"TypeDefOrRef"`
+}
+
+// ExportedType is defined in §II.22.14.
+type ExportedType struct {
+	Flags          flags.TypeAttributes
+	TypeDefID      uint32
```

Comment:
This column is special because it is a hint and because it does not follow the standard `Index` with rules of 2 bytes if the target table has less than 1<<16 rows, else 4 bytes.

I thought that as the ecma-335 authors wanted to make it special, then so do we in order to avoid users using it as a normal index.

Reply:
Will add a comment